builtin/provisioners/chef-client/resource_provisioner.go

```diff
@@ -0,0 +1,402 @@
+func retryFunc(timeout time.Duration, f func() error) error {
+	finish := time.After(timeout)
+	for {
+		err := f()
+		if err == nil {
+			return nil
+		}
+		log.Printf("Retryable error: %v", err)
+
+		select {
+		case <-finish:
+			return err
+		case <-time.After(3 * time.Second):
+		}
+	}
+}
+
+func (p *Provisioner) runChefClientFunc(
+	confDir string) func(terraform.UIOutput, communicator.Communicator) error {
+	return func(o terraform.UIOutput, comm communicator.Communicator) error {
+		fb := path.Join(confDir, firstBoot)
+		cmd := fmt.Sprintf("chef-client -j %q -E %q", fb, p.Environment)
+
+		if p.LogToFile {
+			if err := os.MkdirAll(logfileDir, 0777); err != nil {
+				return fmt.Errorf("Error creating logfile directory %s: %v", logfileDir, err)
+			}
+
+			logFile := path.Join(logfileDir, p.NodeName)
+			f, err := os.Create(path.Join(logFile))
+			if err != nil {
+				return fmt.Errorf("Error creating logfile %s: %v", logFile, err)
+			}
+			f.Close()
+
+			o.Output("Writing Chef Client output to " + logFile)
+			o = p
+		}
+
+		return p.runCommand(o, comm, cmd)
+	}
+}
+
+// Output implementation of terraform.UIOutput interface
+func (p *Provisioner) Output(output string) {
+	logFile := path.Join(logfileDir, p.NodeName)
+	f, err := os.OpenFile(logFile, os.O_APPEND|os.O_WRONLY, 0666)
+	if err != nil {
+		log.Printf("Error creating logfile %s: %v", logFile, err)
+		return
+	}
+	defer f.Close()
+
+	// These steps are needed to remove any ANSI escape codes used to colorize
+	// the output and to make sure we have proper line endings before writing
+	// the string to the logfile.
+	re := regexp.MustCompile(`\x1b\[[0-9;]+m`)
+	output = re.ReplaceAllString(output, "")
+	output = strings.Replace(output, "\r", "\n", -1)
+
+	if _, err := f.WriteString(output); err != nil {
+		log.Printf("Error writing output to logfile %s: %v", logFile, err)
+	}
+
+	if err := f.Sync(); err != nil {
+		log.Printf("Error saving logfile %s to disk: %v", logFile, err)
+	}
+}
+
+func (p *Provisioner) deployConfigFiles(
+	o terraform.UIOutput,
+	comm communicator.Communicator,
+	confDir string) error {
+	// Open the validation .pem file
+	f, err := os.Open(p.ValidationKeyPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Copy the validation .pem to the new instance
+	if err := comm.Upload(path.Join(confDir, "validation.pem"), f); err != nil {
+		return fmt.Errorf("Uploading validation.pem failed: %v", err)
+	}
+
+	// Make strings.Join available for use within the template
+	funcMap := template.FuncMap{
+		"join": strings.Join,
+	}
+
+	// Create a new template and parse the client.rb into it
+	t := template.Must(template.New("client.rb").Funcs(funcMap).Parse(clientConf))
+
+	var buf bytes.Buffer
+	err = t.Execute(&buf, p)
+	if err != nil {
+		return fmt.Errorf("Error executing client.rb template: %s", err)
+	}
+
+	// Copy the client.rb to the new instance
+	if err := comm.Upload(path.Join(confDir, "client.rb"), &buf); err != nil {
```

Comment:
Yeah I think it comes down to the same thing... What use case does that have? Personally I'm not aware of the use case, but if there are some we should look into this one.

Will make it a const in the next commit!